### PR TITLE
Parser optimizations

### PR DESCRIPTION
**Pull request by sand1k**
This pull request is the first one from set of patches aimed to improve frontend of the engine.
In this patch we greatly optimized parser and bytecode generation, reduced memory consumption and parsing time. Here is the short summary of improvements made within this PR:
- Parser has no recursion
- Higher quality of generated bytecode (generates less instructions)
- Peak memory consumption of parser is 2x times lower on merged *.js files from IoTjs
- ~10x times faster parsing of IoTjs
- Full test262 run takes ~17 mins instead of ~35 mins with previous parser. Our upcoming patch is going to reduce this time even more.
- 14 tests from test262 which failed with OUT_OF_MEMORY and OUT_OF_REGISTERS were fixed, due to much lower memory consumption of parser and better bytecode generation
- Fixed native stack usage
- On SunSpider 1.0.2 we got ~12% lower memory consumption and ~10% faster performance with new parser.

This results are not final and will be improved with further patches, which include:
- Performance improvement of literal processing (on test262 ~80% of time is spent in this part)
- Parse-level bytecode optimizations (~30% reduction of generated bytecode)

Detailed measurements for this patch could be found below.

__Peaks for iot.js (measured on RPi2):__
// Master
./jr-old_parser_mem_stats --parse-only --mem-stats merged.js
```
Heap stats:
  Heap size = 260096 bytes
  Chunk size = 64 bytes
  Allocated chunks count = 0
  Allocated = 0 bytes
  Waste = 0 bytes
  Peak allocated chunks count = 3372
  Peak allocated = 212251 bytes
  Peak waste = 3557 bytes
```
// Newest (ec059dd63f63639ea153b8b243d60b4c0e7deb55)
// ./jr-new_parser_mem_stats --parse-only --mem-stats merged.js
```
Heap stats:
  Heap size = 260096 bytes
  Chunk size = 64 bytes
  Allocated chunks count = 0
  Allocated = 0 bytes
  Waste = 0 bytes
  Peak allocated chunks count = 1329
  Peak allocated = 76940 bytes
  Peak waste = 10472 bytes
```
__Parsing time for iot.js (measured on RPi2):__
time ./jr-old_parser --parse-only merged.js

real  0m8.460s
user  0m8.460s
sys 0m0.000s

time ./jr-new_parser --parse-only merged.js

real  0m0.732s
user  0m0.728s
sys 0m0.000s

Improvement: +91%

__Sunspider results__

Benchmark| RSS | Perf
------------ | -------------|--------------
                                ./sunspider-1.0.2/3d-cube.js |    128 ->    112 (12.500%) | 1.94833 ->   1.61 (17.365%) | 
                    ./sunspider-1.0.2/access-binary-trees.js |     84 ->     84 (0.000%) | 1.03667 -> 1.00167 (3.376%) | 
                        ./sunspider-1.0.2/access-fannkuch.js |     48 ->     40 (16.667%) |  4.895 -> 4.81167 (1.702%) | 
                           ./sunspider-1.0.2/access-nbody.js |     60 ->     48 (20.000%) |   2.42 -> 2.22667 (7.989%) | 
               ./sunspider-1.0.2/bitops-3bit-bits-in-byte.js |     36 ->     32 (11.111%) | 1.29667 -> 1.13333 (12.597%) | 
                    ./sunspider-1.0.2/bitops-bits-in-byte.js |     36 ->     32 (11.111%) |  1.755 -> 1.45667 (16.999%) | 
                     ./sunspider-1.0.2/bitops-bitwise-and.js |     32 ->     32 (0.000%) | 1.85167 ->   1.63 (11.971%) | 
                  ./sunspider-1.0.2/controlflow-recursive.js |    216 ->    220 (-1.852%) | 0.858333 -> 0.786667 (8.349%) | 
                            ./sunspider-1.0.2/crypto-sha1.js |    140 ->    132 (5.714%) |   8.24 -> 8.52167 (-3.418%) | 
                      ./sunspider-1.0.2/date-format-xparb.js |     92 ->     72 (21.739%) | 1.23667 -> 1.00667 (18.598%) | 
                            ./sunspider-1.0.2/math-cordic.js |     44 ->     40 (9.091%) |  1.905 -> 1.63167 (14.348%) | 
                      ./sunspider-1.0.2/math-partial-sums.js |     44 ->     32 (27.273%) | 1.12667 -> 0.953333 (15.385%) | 
                     ./sunspider-1.0.2/math-spectral-norm.js |     48 ->     40 (16.667%) |   1.13 ->   1.09 (3.540%) | 
                           ./sunspider-1.0.2/string-fasta.js |     56 ->     48 (14.286%) |  3.235 -> 3.01667 (6.749%) | 
                                              Geometric mean |                         12.1303% |                          9.9108%

__Test262 runnting time__
Old parser 35 min
New parser 17 min

__Number of instructions__
Iot.js:
Old parser: 10170
New parser: 9628

Test262 header files (cth.js sta.js ed.js testBuiltInObject.js testIntl.js):
Old parser: 6542
New parser: 6063

__Fixed tests from Test262__
S11.7.1_A4_T1 
S11.7.1_A4_T2 
S11.7.1_A4_T3 
S11.7.1_A4_T4
S11.7.2_A4_T1 
S11.7.2_A4_T2 
S11.7.2_A4_T3 
S11.7.2_A4_T4 
S11.7.3_A4_T1 
S11.7.3_A4_T2 
S11.7.3_A4_T3 
S11.7.3_A4_T4 
S12.6.4_A5.1 
S12.6.4_A5

### CODE REVIEW

**Comment by egavrin** [comment]
Add related issues: #737, #646, #644

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+  bc_header_p->is_ref_arguments_identifier = is_ref_arguments_identifier;
+  bc_header_p->is_ref_eval_identifier = is_ref_eval_identifier;
+  bc_header_p->is_args_moved_to_regs = is_arguments_moved_to_regs;
+  bc_header_p->is_no_lex_env = is_no_lex_env;
+} /* bc_add_bytecode_data */
+
+static void
+bc_free_bytecode_data (bytecode_data_header_t *bytecode_data_p)
+{
+  bytecode_data_header_t* next_to_handle_list_p = bytecode_data_p;
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: `*`

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+  }
+} /* bc_free_bytecode_data */
+
+/**
+ * Deletes bytecode and associated hash table
+ */
+void
+bc_remove_bytecode_data (const bytecode_data_header_t *bytecode_data_p)
+{
+  bytecode_data_header_t *prev_header = NULL;
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: `_p`

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+                                                       MEM_ALIGNMENT);
+  const size_t header_and_tables_size = JERRY_ALIGNUP ((sizeof (bytecode_data_header_t)
+                                                        + hash_table_size
+                                                        + declarations_area_size),
+                                                       MEM_ALIGNMENT);
+
+  uint8_t *buffer_p = (uint8_t *) mem_heap_alloc_block (bytecode_size + header_and_tables_size,
+                                                        MEM_HEAP_ALLOC_LONG_TERM);
+
+  lit_id_hash_table *lit_id_hash = lit_id_hash_table_init (buffer_p + sizeof (bytecode_data_header_t),
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: `_p`

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+                        scope_p->is_args_moved_to_regs,
+                        scope_p->is_no_lex_env);
+
+  JERRY_ASSERT (scope_p->bc_header_cp == MEM_CP_NULL);
+  MEM_CP_SET_NON_NULL_POINTER (scope_p->bc_header_cp, header_p);
+
+  return header_p;
+} /* bc_dump_single_scope */
+
+static bytecode_data_header_t*
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: ` *`

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+ *          Note to declarations_cp field of bytecode_data_header_t
+ *
+ * @return pointer to bytecode header of the outer most scope
+ */
+bytecode_data_header_t *
+bc_alloc_headers_from_scopes_tree (scopes_tree scope_p)
+{
+  bytecode_data_header_t *bc_header_p = bc_dump_scope_and_prepare_header (scope_p);
+
+  bytecode_data_header_t* next_to_handle_list_p = bc_header_p;
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: `*`

File: jerry-core/parser/js/scopes-tree.cpp [hunk 6]
```diff
@@ -175,582 +96,22 @@ scopes_tree_variable_declaration_exists (scopes_tree tree, /**< scope */
...
-   Init literal indexes 'hash' table.
-   Reorder function declarations.
-   Rewrite instructions' temporary uids with their keys in literal indexes 'hash' table. */
-vm_instr_t *
-scopes_tree_raw_data (scopes_tree tree, /**< scopes tree to convert to byte-code array */
-                      uint8_t *buffer_p, /**< buffer for byte-code array and literal identifiers hash table */
-                      size_t instructions_array_size, /**< size of space for byte-code array */
-                      lit_id_hash_table *lit_ids) /**< literal identifiers hash table */
+void
+scopes_tree_dump_var_decls (scopes_tree scope_p, /**< scopes tree */
```
Comment (egavrin) [inline jerry-core/parser/js/scopes-tree.cpp]: Remove `_p`

File: jerry-core/vm/opcodes.cpp [hunk 7]
```diff
@@ -17,6 +17,7 @@
 #include "jrt.h"
 #include "opcodes.h"
 #include "opcodes-ecma-support.h"
+#include "bytecode-data.h"
```
Comment (egavrin) [inline jerry-core/vm/opcodes.cpp]: Reorder includes

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+ */
+
+#include "bytecode-data.h"
+#include "pretty-printer.h"
+#include "opcodes-dumper.h"
+
+static bytecode_data_header_t *first_bytecode_header_p = NULL;
+
+static void
+bc_add_bytecode_data (bytecode_data_header_t *bc_header_p,
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Missing comments

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+
+  bc_header_p->is_strict = is_strict;
+  bc_header_p->is_ref_arguments_identifier = is_ref_arguments_identifier;
+  bc_header_p->is_ref_eval_identifier = is_ref_eval_identifier;
+  bc_header_p->is_args_moved_to_regs = is_arguments_moved_to_regs;
+  bc_header_p->is_no_lex_env = is_no_lex_env;
+} /* bc_add_bytecode_data */
+
+static void
+bc_free_bytecode_data (bytecode_data_header_t *bytecode_data_p)
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Missing comments

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+
+    JERRY_ASSERT (bc_header_list_iter_p == NULL);
+  }
+} /* bc_free_bytecode_data */
+
+/**
+ * Deletes bytecode and associated hash table
+ */
+void
+bc_remove_bytecode_data (const bytecode_data_header_t *bytecode_data_p)
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Missing comments

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+
+      break;
+    }
+
+    prev_header = cur_header_p;
+    cur_header_p = MEM_CP_GET_POINTER (bytecode_data_header_t, cur_header_p->next_header_cp);
+  }
+} /* bc_remove_bytecode_data */
+
+vm_instr_t bc_get_instr (const bytecode_data_header_t *bytecode_data_p,
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Missing comments

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+                         vm_instr_counter_t oc)
+{
+  JERRY_ASSERT (oc < bytecode_data_p->instrs_count);
+  return bytecode_data_p->instrs_p[oc];
+}
+
+void
+bc_print_instrs (const bytecode_data_header_t *bytecode_data_p)
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Missing comments

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+#endif
+}
+
+/**
+ * Dump single scopes tree into bytecode
+ *
+ * @return pointer to bytecode header of the outer most scope
+ */
+static bytecode_data_header_t *
+bc_dump_single_scope (scopes_tree scope_p)
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Missing comments

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+                        scope_p->is_no_lex_env);
+
+  JERRY_ASSERT (scope_p->bc_header_cp == MEM_CP_NULL);
+  MEM_CP_SET_NON_NULL_POINTER (scope_p->bc_header_cp, header_p);
+
+  return header_p;
+} /* bc_dump_single_scope */
+
+static bytecode_data_header_t*
+bc_dump_scope_and_prepare_header (scopes_tree scope_p)
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Missing comments

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 15]
```diff
@@ -0,0 +1,488 @@
...
+/**
+ * Dump scopes tree into bytecode
+ *
+ * See also:
+ *          Note to declarations_cp field of bytecode_data_header_t
+ *
+ * @return pointer to bytecode header of the outer most scope
+ */
+bytecode_data_header_t *
+bc_alloc_headers_from_scopes_tree (scopes_tree scope_p)
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Missing comments

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 16]
```diff
@@ -0,0 +1,488 @@
...
+  }
+
+  MEM_CP_SET_POINTER (bc_header_p->next_header_cp, first_bytecode_header_p);
+  first_bytecode_header_p = bc_header_p;
+
+  return bc_header_p;
+} /* bc_alloc_headers_from_scopes_tree */
+
+void
+bc_finalize ()
```
Comment (egavrin) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Missing comments

File: jerry-core/parser/js/jsp-early-error.cpp [hunk 17]
```diff
@@ -195,19 +189,7 @@ jsp_early_error_check_for_duplication_of_prop_names (bool is_strict, locus loc _
...
-
-void jsp_early_error_add_varg (jsp_operand_t op)
-{
-  JERRY_ASSERT (op.is_literal_operand ());
-  STACK_PUSH (props, create_prop_literal (lit_get_literal_by_cp (op.get_literal ()), VARG));
-}
-
-static void
-emit_error_on_eval_and_arguments (literal_t lit, locus loc __attr_unused___)
+jsp_early_error_emit_error_on_eval_and_arguments (literal_t lit, locus loc)
```
Comment (egavrin) [inline jerry-core/parser/js/jsp-early-error.cpp]: Missing comments

File: jerry-core/parser/js/lexer.cpp [hunk 18]
```diff
@@ -1262,6 +1265,59 @@ lexer_parse_comment (void)
   return false;
 } /* lexer_parse_comment */
 
+static bool
+lexer_skip_whitespace_and_comments (void)
```
Comment (egavrin) [inline jerry-core/parser/js/lexer.cpp]: Missing comments

File: jerry-core/parser/js/lexer.cpp [hunk 19]
```diff
@@ -1519,8 +1535,11 @@ lexer_parse_token (bool maybe_regexp) /**< read '/' as regexp? */
 } /* lexer_parse_token */
 
 token
-lexer_next_token (bool maybe_regexp) /**< read '/' as regexp? */
+lexer_next_token (bool maybe_regexp, /**< read '/' as regexp? */
```
Comment (egavrin) [inline jerry-core/parser/js/lexer.cpp]: Missing comments

File: jerry-core/parser/js/lexer.cpp [hunk 20]
```diff
@@ -1809,15 +1755,73 @@ lexer_token_type_to_string (token_type tt)
...
+    case TOK_KW_WITH: return "with";
+    case TOK_KW_YIELD: return "yield";
     default: JERRY_UNREACHABLE ();
   }
 }
 
-void
-lexer_set_strict_mode (bool is_strict)
+jsp_token_type_t __attr_always_inline___
+lexer_get_token_type (token t)
```
Comment (egavrin) [inline jerry-core/parser/js/lexer.cpp]: Missing comments

File: jerry-core/parser/js/scopes-tree.cpp [hunk 21]
```diff
@@ -829,6 +190,23 @@ scopes_tree_strict_mode (scopes_tree tree)
   return (bool) tree->strict_mode;
 }
 
+size_t
+scopes_tree_child_scopes_num (scopes_tree tree)
```
Comment (egavrin) [inline jerry-core/parser/js/scopes-tree.cpp]: Missing comments

**Comment by egavrin** [comment]
@zherczeg @galpeter @LaszloLango could you please review this PR. This PR brings great enhancements to parser and I'd like to merge it to master this week - without it we cannot continue our planned activities.

**Comment by egavrin** [comment]
@seanshpark could you please verify iotjs with new parser and snapshot from this pr?

**Comment by seanshpark** [comment]
I think @chunseoklee would be better to check :)

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 22]
```diff
@@ -0,0 +1,865 @@
...
+  uint8_t is_ref_arguments_identifier : 1; /**< code doesn't reference 'arguments' identifier */
+  uint8_t is_ref_eval_identifier : 1; /**< code doesn't reference 'eval' identifier */
+  uint8_t is_args_moved_to_regs : 1; /**< the function's arguments are moved to registers,
+                                      *   so should be initialized in vm registers,
+                                      *   and not in lexical environment */
+  uint8_t is_no_lex_env : 1; /**< no lex. env. is necessary for the scope */
+} jerry_snapshot_bytecode_header_t;
+
+static void
+bc_add_bytecode_data (bytecode_data_header_t *bc_header_p, /**< byte-code scope data header */
```
Comment (zherczeg) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Usually there is some description before functions.

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 23]
```diff
@@ -0,0 +1,865 @@
...
+
+    while (bc_header_list_iter_p != NULL)
+    {
+      bytecode_data_header_t *header_p = bc_header_list_iter_p;
+
+      bc_header_list_iter_p = MEM_CP_GET_POINTER (bytecode_data_header_t, header_p->next_header_cp);
+
+      mem_cpointer_t *declarations_p = MEM_CP_GET_POINTER (mem_cpointer_t, header_p->declarations_cp);
+
+      for (uint32_t index = 0; index < header_p->func_scopes_count; index++)
```
Comment (zherczeg) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: we should avoid C++ in new code

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 24]
```diff
@@ -0,0 +1,865 @@
...
+    return INVALID_LITERAL;
+  }
+
+  return lit_id_hash_table_lookup (lit_id_hash, id, oc);
+} /* bc_get_literal_cp_by_uid */
+
+
+
+
+#ifdef JERRY_ENABLE_SNAPSHOT
```
Comment (zherczeg) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: Too many newlines.

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 25]
```diff
@@ -0,0 +1,865 @@
...
+  bytecode_header.func_scopes_count = bytecode_data_p->func_scopes_count;
+  bytecode_header.var_decls_count = bytecode_data_p->var_decls_count;
+  bytecode_header.is_strict = bytecode_data_p->is_strict;
+  bytecode_header.is_ref_arguments_identifier = bytecode_data_p->is_ref_arguments_identifier;
+  bytecode_header.is_ref_eval_identifier = bytecode_data_p->is_ref_eval_identifier;
+  bytecode_header.is_args_moved_to_regs = bytecode_data_p->is_args_moved_to_regs;
+  bytecode_header.is_no_lex_env = bytecode_data_p->is_no_lex_env;
+  size_t bytecode_header_offset = *in_out_buffer_offset_p;
+
+  // Dump instructions
```
Comment (zherczeg) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: // comment?

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 26]
```diff
@@ -0,0 +1,865 @@
...
+                       const bytecode_data_header_t *bytecode_data_p, /**< byte-code data */
+                       const lit_mem_to_snapshot_id_map_entry_t *lit_map_p, /**< map from literal
+                                                                             *   identifiers in
+                                                                             *   literal storage
+                                                                             *   to literal offsets
+                                                                             *   in snapshot */
+                       uint32_t literals_num, /**< literals number */
+                       uint32_t *out_scopes_num) /**< number of scopes */
+{
+  bytecode_data_header_t* next_to_handle_list_p = first_bytecode_header_p;
```
Comment (zherczeg) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: *

File: jerry-core/parser/js/bc/bytecode-data.cpp [hunk 27]
```diff
@@ -0,0 +1,865 @@
...
+  {
+    instrs_p = snapshot_instrs_p;
+  }
+
+  buffer_offset += bytecode_header.instrs_size; // buffer_offset is now offset of variable declarations
+
+  // Read uid->lit_cp hash table
+  uint8_t *lit_id_hash_table_buffer_p = buffer_p + sizeof (bytecode_data_header_t);
+  if (!(lit_id_hash_table_load_from_snapshot (blocks_count,
+                                             idx_num_total,
```
Comment (zherczeg) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: alignment?

**Comment by zherczeg** [comment]
There are lots of style issues, but you are aware of it. Since a lot of code moved around it is hard to judge all changes, but likely they are ok.

**Comment by seanshpark** [comment]
IoT.js unit test are all OK with and without snapshot. IoT.js needed some changes to reflect Jerry changes.

**Comment by egavrin** [comment]
@seanshpark great! Thank you.

**Comment by seanshpark** [comment]
Just curious, `Peak waste = 3557 bytes` became `Peak waste = 17900 bytes`. As I don't deeply understand the `waste` value, is this something we don't have to worry about or something to fix in near future?

**Comment by egavrin** [comment]
@seanshpark we should care about waste, but it's a bit less important than peak allocated for now. We studied the reasons and going to improve it.

**Comment by akiss77** [comment]
@egavrin What is waste after all? I've seen this question raised at #630 earlier but with no explanation eventually.

**Comment by egavrin** [comment]
@akiss77 waste indicated internal fragmentation caused by inefficient allocation — memory wasted because the memory requested is a little smaller than a block size. Because of the way current memory allocation technique works, a program that requests 40 bytes of memory would be allocated 64 bytes, which results in a waste of 24 bytes of memory.

**Comment by akiss77** [comment]
@egavrin Thanks for the clarification!

File: jerry-core/parser/js/bc/bytecode-data.cpp (hunk 23 above)
Comment (sand1k) [inline jerry-core/parser/js/bc/bytecode-data.cpp]: As was discussed in the chat, C99 supports for (int i...) way of variables declaration.

**Comment by ruben-ayrapetyan** [comment]
Measurements on Sunspider and IoTjs peaks were updated to latest commit.

**Comment by egavrin** [comment]
@zherczeg @galpeter @LaszloLango we updated the PR:
- Reduced peak memory consumption 110k -> 77k
- Reduced waste 17k -> 10k
- Fixed all found style issues and added lots of comment across the code

Please take a look, can we merge it?

**Comment by zherczeg** [comment]
LGTM
Nice job!

**Comment by ruben-ayrapetyan** [comment]
@zherczeg, thank you for review!

**Comment by egavrin** [comment]
LGTM

**Pull request merged**